streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -46,54 +46,60 @@
-                            V value,
-                            Integer partition,
-                            Long timestamp,
-                            Serializer<K> keySerializer,
-                            Serializer<V> valueSerializer) {
-        send(topic, key, value, partition, timestamp, keySerializer, valueSerializer, null);
+                            final K key,
+                            final V value,
+                            final Long timestamp,
+                            final Serializer<K> keySerializer,
+                            final Serializer<V> valueSerializer,
+                            final StreamPartitioner<? super K, ? super V> partitioner) {
+        Integer partition = null;
+
+        if (partitioner != null) {
+            final List<PartitionInfo> partitions = producer.partitionsFor(topic);
+            if (partitions.size() > 0) {
+                partition = partitioner.partition(key, value, partitions.size());
+            } else {
+                throw new StreamsException("Could not get partition information for topic '" + topic + "'." +
+                    " This can happen if the topic does not exist.");
+            }
+        }
+
+        send(topic, key, value, partition, timestamp, keySerializer, valueSerializer);
     }
 
     @Override
     public <K, V> void  send(final String topic,
-                             K key,
-                             V value,
-                             Integer partition,
-                             Long timestamp,
-                             Serializer<K> keySerializer,
-                             Serializer<V> valueSerializer,
-                             StreamPartitioner<? super K, ? super V> partitioner) {
+                             final K key,
+                             final V value,
+                             final Integer partition,
+                             final Long timestamp,
+                             final Serializer<K> keySerializer,
+                             final Serializer<V> valueSerializer) {
         checkForException();
-        byte[] keyBytes = keySerializer.serialize(topic, key);
-        byte[] valBytes = valueSerializer.serialize(topic, value);
-        if (partition == null && partitioner != null) {
-            List<PartitionInfo> partitions = this.producer.partitionsFor(topic);
-            if (partitions != null && partitions.size() > 0)
-                partition = partitioner.partition(key, value, partitions.size());
-        }
+        final byte[] keyBytes = keySerializer.serialize(topic, key);
+        final byte[] valBytes = valueSerializer.serialize(topic, value);
 
-        ProducerRecord<byte[], byte[]> serializedRecord =
+        final ProducerRecord<byte[], byte[]> serializedRecord =
                 new ProducerRecord<>(topic, partition, timestamp, keyBytes, valBytes);
 
         for (int attempt = 1; attempt <= MAX_SEND_ATTEMPTS; attempt++) {
```

Comment:
I only added a comment -- IMHO, it's better to start counting from 1 for this specific case.